runtime/cudaq/utils/tensor.cpp

```diff
@@ -0,0 +1,102 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "cudaq/utils/tensor.h"
+#include <strstream>
+
+inline std::complex<double> &access(std::complex<double> *p,
+                                    cudaq::matrix_2::Dimensions sizes,
+                                    std::size_t row, std::size_t col) {
+  return p[row * sizes.second + col];
+}
+
+cudaq::matrix_2 &cudaq::matrix_2::operator*=(const cudaq::matrix_2 &right) {
+  if (get_columns() != right.get_rows()) {
+    clear();
+    return *this;
+  }
+  auto new_data = new std::complex<double>[get_rows() * right.get_columns()];
+  for (std::size_t i = 0; i < get_rows(); i++)
```

Comment:
do we need to check for the dimensions to match?

Reply:
Yes. That's done on line 19.

BTW, errors like that are "reported" by returning an empty matrix. Feel free to suggest something more exciting.

Reply:
I suggest throwing an error